integration/util.py

```diff
@@ -37,17 +38,29 @@ class _CollectOutputProtocol(ProcessProtocol):
     self.output, and callback's on done with all of it after the
     process exits (for any reason).
     """
-    def __init__(self):
+    def __init__(self, stdin=None):
         self.done = Deferred()
         self.output = StringIO()
+        self._stdin = stdin
+
+    def connectionMade(self):
+        if self._stdin is not None:
+            self.transport.write(self._stdin)
+            self.transport.closeStdin()
 
     def processEnded(self, reason):
         if not self.done.called:
             self.done.callback(self.output.getvalue())
 
     def processExited(self, reason):
         if not isinstance(reason.value, ProcessDone):
-            self.done.errback(reason)
+            #self.done.errback(reason)
+            self.done.errback(RuntimeError(
+                "Process failed: {}\nOutput:\n{}".format(
+                    reason,
+                    self.output.getvalue(),
+                )
+            ))
```

Comment:
`reason` will always be `ProcessTerminated` won't it?

So certainly this could be something besides `RuntimeError` and pass the actual ProcessTerminated instance (I guess in case you want to look at `.exitCode`?). This could include `output` instead of `output.getvalue()` .. but what would a caller do besides look at the stuff in `.getvalue()`?

Reply:
Added a custom exception